### DISCUSSION TITLE
User and Admin Functionality Not Working Together-- Issue in logic within App.js useEffect()?

### DISCUSSION THREAD

**Discussion by thedigitalmenagerie** [nss-evening-cohort-14/evening-client-side#256] [Help Me!] [answered]
Honey-Rae Swan
Have you searched the other help tickets? Yes, there are no similar issue tickets. 

Describe your issue

With the current logic, user functionality is working and it does not reach admin. Admin is console.warning as null from the beginning in App.js. I think this has to do with the various attempts/ reworkings of the logic with the if statements in useEffect() not being quite right. 

List any error messages that appear

No error messages in the current form-- other attempts to fix the issue were just creating errors when I attempted two useEffect() statements.

What have you googled? 

I am not really sure what to google, we went through this as a study group tonight and I attempted the fix Dr. T suggested but I think I am having an issue with combining the logic of user and admin together. 

What have you tried? 

The initial issue was when I attempted to add user to the if statements. Aja helped me pass the entire object on BioView as opposed to each key value pair and helped me catch I was simply forgetting to pass user to the component and the route. I was logged into my non-admin account, and buttons disappeared. After refactoring all other pages based on what Aja went over with me Tuesday, I checked admin functionality yesterday (too late) and it wasn't working. 

1. Using two useEffects in App.js (errors galore)
2. I tried the four part if statement and set user and Admin in each of them. 
3. I tried to reflect the three states of user and set user and admin in each of them.
4. All of the below examples in reverse order with admin statements before user:

Example 1 (original-- does not show admin CRUD anywhere):

  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed) {
        const userInfoObj = {
          fullName: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          user: authed.email.split('@')[0]
        };
        setUser(userInfoObj);
      } else if (user || user === null) {
        setUser(false);
      } else if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
      } else if (admin || admin === null) {
        setAdmin(false);
      }
    });
  }, []);

Example 2 (setting both user and admin in each of the parts in a four part if statement, tried prior to bringing to group-- shows full admin crud whether a user is logged in or not):
  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed) {
        const userInfoObj = {
          fullName: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          user: authed.email.split('@')[0]
        };
        setUser(userInfoObj);
        setAdmin(false);
      } else if (user || user === null) {
        setUser(false);
        setAdmin(false);
      } else if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
        setUser(false);
      } else if (admin || admin === null) {
        setAdmin(false);
        setUser(false);
      }
    });
  }, []);

Example 3 (using the 3 part if statement, based on discussion with study group -- shows admin CRUD when there is no user and no admin and in logged in view of both user and admin):

  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed) {
        const userInfoObj = {
          fullName: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          user: authed.email.split('@')[0]
        };
        setUser(userInfoObj);
        setAdmin(false);
      } else if ((user || user === null) && (admin || admin === null)) {
        setUser(false);
        setAdmin(false);
      } else if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
        setUser(false);
      }
    });
  }, []);

Example 4 (same as above-- but with || instead of &&-- shows admin crud whether someone is logged in or not):
  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed) {
        const userInfoObj = {
          fullName: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          user: authed.email.split('@')[0]
        };
        setUser(userInfoObj);
        setAdmin(false);
      } else if ((user || user === null) || (admin || admin === null)) {
        setUser(false);
        setAdmin(false);
      } else if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
        setUser(false);
      }
    });
  }, []);

Copy/paste link TO THE FILE on your branch or in your repo that you are having issues with NO SCREEN SHOTS 🥰

https://github.com/thedigitalmenagerie/hr-react-personal-bio/blob/aafunctionality/src/App/App.js

**Comment by drteresavasquez** [comment]
You need to put the admin block first. If the first condition is true, which it is, it will run and break out of the if else.

**Reply by thedigitalmenagerie** [reply]
This is what I tried:

useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed && (authed.uid === process.env.REACT_APP_ADMIN_UID)) {
        setAdmin(true);
        setUser(false);
      } else if (authed) {
        const userInfoObj = {
          fullName: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          user: authed.email.split('@')[0]
        };
        setUser(userInfoObj);
        setAdmin(false);
      } else if ((user || user === null) || (admin || admin === null)) {
        setUser(false);
        setAdmin(false);
      }
    });
  }, []);

This version shows all the buttons regardless of whether anyone is logged in or not. It makes total sense to me why it would be set up this way and what it should be doing-- am I missing something with this attempt/ not following what you said above correctly?

**Reply by thedigitalmenagerie** [reply]
I also did this with ((user || user === null) && (admin || admin === null)) in the final else if.

**Reply by drteresavasquez** [reply]
I would say simplify and drop the extra user until you get the admin working. Then try the extra user again.

**Comment by drteresavasquez** [comment]
@thedigitalmenagerie were you able to get this working doing the simplified version?

**Reply by thedigitalmenagerie** [reply]
Had to go to bed last night-- works starts so EARLY! Lol.  I am starting work on this right now on a new branch (while working-- so, might take me a little longer today). I will get back to this and let you know if I got it working by removing user and getting admin going again before attempting to add user back.

**Reply by drteresavasquez** [reply]
Perfect. Thank you and keep us posted. I know all about time so I just wanted to follow up. If you get it working with admin, I would stay that course and try adding another user after you e completed the project. I believe Chie also completed that so reach out to her if you'd like to see what she did as well.

**Reply by thedigitalmenagerie** [reply]
Will do! I thought admin was working the whole time because I was testing on my non-admin account for too long before checking back on admin. Rookie mistake to not test everything at every step!!! So, the rest of my project is at 100% of what I needed/ wanted and this is the only issue, thankfully. Haha.

**Reply by thedigitalmenagerie** [reply]
Looks like I needed to pass logic that better reflected my three states through the app! I will update with a detailed answer when I finish passing everything but, for now, I am doing a happy dance and REACTING!!

**Reply by drteresavasquez** [reply]
Yassss!!!!! 🤓🎉

**Reply by thedigitalmenagerie** [reply]
Everything is working-- but one thing!! I have tested all three states but for the contact and reviews page (which are private routes), I have to switch between these two versions of the Private Routes function: 

The first version allows me to see the non-admin state in private routes and redirects to home in admin state when viewing the contact or reviews page:

const PrivateRoute = ({
  component: Component,
  user,
  admin,
  ...rest
}) => {
  const routeChecker = (taco) => (admin
    ? (<Component {...taco} admin={admin}/>)
    : (<Redirect to={{ pathname: '/', state: { from: taco.location } }} />));
  const routeChecker2 = (taco) => (user
    ? (<Component {...taco} user={user}/>)
    : (<Redirect to={{ pathname: '/', state: { from: taco.location } }} />));
  return <Route {...rest} render={(props) => (routeChecker(props) **&&** routeChecker2(props))} />;
};

The second version allows me to see the admin state in private routes, and redirects to home in non-admin state when viewing the contact or reviews page:

const PrivateRoute = ({
  component: Component,
  user,
  admin,
  ...rest
}) => {
  const routeChecker = (taco) => (admin
    ? (<Component {...taco} admin={admin}/>)
    : (<Redirect to={{ pathname: '/', state: { from: taco.location } }} />));
  const routeChecker2 = (taco) => (user
    ? (<Component {...taco} user={user}/>)
    : (<Redirect to={{ pathname: '/', state: { from: taco.location } }} />));
  return <Route {...rest} render={(props) => (routeChecker(props) **||** routeChecker2(props))} />;
};

I am going to work with modifying this a little longer but would appreciate any help you think is appropriate!

**Reply by thedigitalmenagerie** [reply]
Forgot the new branch:

https://github.com/thedigitalmenagerie/hr-react-personal-bio/tree/removingUser

**Reply by drteresavasquez** [reply]
You can have 2 types of private routes. In reality you only want the private routes to show if the user is admin, right? 

What are you hoping to achieve for the other user? Are they supposed to see anything different? If not your admin stuff should be fine.

**Reply by thedigitalmenagerie** [reply]
I wanted the private routes to show for both user and admin but have the page display different content based on whether it was user or admin (contact form/review form for user and contact/review cards only for admin). It works fine with both the forms and cards displaying for admin. So, I could re-work things to do that.

**Reply by drteresavasquez** [reply]
Oh ok. Then make sure to check for whichever user type is logged in and make different components for each user type. In your return statement in the view, do something like you did for the joke generator where if the user is admin show the component you made for admin and same for logged in user.

**Comment by GonzalesMatthew** [comment]
I had a similar issue when signing out as a user because I realized the logic was jumping out ahead of when it checked for my user status ... so I had to combine both the user and admin status check into one 'else-if' code block. I don't think it's quite the same issue you are having but similar and here is my code block for an example on how working logic looks if you use both 'user' and 'admin' in your project
```
useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed && (authed.uid === firebaseConfig.adminId)) {
        setAdmin(true);
      } else if (authed) {
        const userInfoObj = {
          fullname: authed.displayName,
          profileImage: authed.photoURL,
          uid: authed.uid,
          email: authed.email
        };
        checkUser(authed.uid).then((response) => {
          if (Object.values(response.data).length === 0) {
            createUser(userInfoObj).then((resp) => setUser(resp));
          } else {
            setUser(userInfoObj);
          }
        });
      } else if ((admin || admin === null) || (user || user === null)) {
        setAdmin(false);
        setUser(false);
      }
    });
  }, []);
```

**Comment by thedigitalmenagerie** [comment, marked as answer]
Routes are for URL's and logic on who sees what is for Views!

First, I needed to put the admin block first. I was creating an if statement where the first condition was true and it would run and  break out of the if else. 

Second, I needed logic that allowed user or admin to access the private routes. This included passing both the user and admin to my private Route component that Dr. T shared with us, which I was doing, but needed to include || instead of the comma that was giving me errors and the subsequent "fixes."

Third, I needed the navbar to show the private routes for both user and admin, this part was written correctly.

Fourth, I needed to put the logic for what the user sees (and admin doesn't) and what the admin sees (and user doesn't) on the same private route they both have access to. This part was written correctly.

In a nutshell, I was adding additional logic and creating two private routes (one with user and one with admin) when I only needed one private route with user and admin access. The logic for what the user sees and what the admin sees belonged only in the view they both had access to.
